lxd/instance/drivers/driver_qemu.go

```diff
@@ -3028,16 +3171,78 @@ func (d *qemu) addDriveConfig(sb *strings.Builder, fdFiles *[]*os.File, bootInde
 		d.devPaths = append(d.devPaths, srcDevPath)
 	}
 
-	return qemuDrive.Execute(sb, map[string]interface{}{
-		"devName":   driveConf.DevName,
-		"devPath":   driveConf.DevPath,
-		"bootIndex": bootIndexes[driveConf.DevName],
-		"cacheMode": cacheMode,
-		"aioMode":   aioMode,
-		"media":     media,
-		"shared":    driveConf.TargetPath != "/" && !strings.HasPrefix(driveConf.DevPath, "rbd:"),
-		"readonly":  shared.StringInSlice("ro", driveConf.Opts),
-	})
+	// The root drive cannot added using qmp due to permission issues.
+	// It therefore is the only block drive which will be added to the qemu.conf file.
+	if driveConf.TargetPath == "/" {
+		return nil, qemuDrive.Execute(sb, map[string]interface{}{
+			"devName":   driveConf.DevName,
+			"devPath":   driveConf.DevPath,
+			"bootIndex": bootIndexes[driveConf.DevName],
+			"cacheMode": cacheMode,
+			"aioMode":   aioMode,
+			"media":     media,
+			"shared":    false,
+			"readonly":  shared.StringInSlice("ro", driveConf.Opts),
+		})
+	}
+
+	// QMP uses two separate values for the cache.
+	directCache := true   // Bypass host cache, use O_DIRECT semantics by default.
+	noFlushCache := false // Don't ignore any flush requests for the device.
+
+	if cacheMode == "unsafe" {
+		directCache = false
+		noFlushCache = true
+	} else if cacheMode == "writeback" {
+		directCache = false
+	}
+
+	blockDev := map[string]interface{}{
+		"aio": aioMode,
+		"cache": map[string]interface{}{
+			"direct":   directCache,
+			"no-flush": noFlushCache,
+		},
+		"discard":   "unmap",
+		"driver":    "file",
+		"filename":  driveConf.DevPath,
+		"node-name": fmt.Sprintf("%s%s", qemuBlockDevIDPrefix, driveConf.DevName), // Node names may only be 31 characters long.
+		"read-only": false,
+	}
+
+	if shared.StringInSlice("ro", driveConf.Opts) {
+		blockDev["read-only"] = "true"
+	}
+
+	if !strings.HasPrefix(driveConf.DevPath, "rbd:") {
+		blockDev["locking"] = "off"
+	}
+
+	device := map[string]string{
+		"id":        fmt.Sprintf("%s%s", qemuDeviceIDPrefix, driveConf.DevName),
+		"bootindex": strconv.Itoa(bootIndexes[driveConf.DevName]),
+		"drive":     blockDev["node-name"].(string),
+		"bus":       "qemu_scsi.0",
+		"channel":   "0",
+		"lun":       "1",
+	}
+
+	if media == "disk" {
+		device["driver"] = "scsi-hd"
+	} else if media == "cdrom" {
+		device["driver"] = "scsi-cd"
+	}
+
+	monHook := func(m *qmp.Monitor) error {
+		err := m.AddBlockDevice(blockDev, device)
```

Comment:
Is there a reason why the call to `AddBlockDevice` in `deviceAttachBlockDevice` is missing a lot of the caching and other logic from this function? Would it make sense to centralize this logic so its used for both start time and hot-plug time?

Reply:
As stated above, this should be `monHook, err := d.addDriveConfig(...)`.